File: Packages/com.tumbangpreso.core/Runtime/Social.cs
using System;
using System.Collections.Generic;

namespace TumbangPreso.Core
{
    /// <summary>
    /// Where a player is, as far as anybody else needs to know.
    ///
    /// ⚠️⚠️ FIVE STATES AND NOT ONE MORE, BECAUSE THE ONE THING ON A FRIENDS LIST IS **WHO IS
    /// ONLINE NOW** (`FUTURE.md` § 0.5b's per-phase table). Every state past that is a submenu,
    /// and a list that reports six shades of busy is a list nobody reads. `FUTURE.md` § 6 names
    /// exactly these: online, in menu, in queue, in a match, spectating — with "online" and "in
    /// menu" collapsed here because they are the same fact told twice.
    ///
    /// ⚠️ `Queued` IS CARRIED THOUGH NOTHING SETS IT YET. Phase 7 owns the queue. A presence
    /// vocabulary that has to grow on the wire later is a vocabulary two builds disagree about,
    /// and `Roster.Slippers` records what inserting a value into a replicated list costs. The
    /// value is here, unused, and says so.
    /// </summary>
    public enum PresenceState
    {
        Offline = 0,
        Menu = 1,
        Queued = 2,
        InMatch = 3,
        Spectating = 4,
    }

    /// <summary>
    /// One person on your list, and everything a rail needs to draw them.
    ///
    /// ⚠️⚠️ THE PLAYER ID IS THE IDENTITY AND THE HANDLE IS A LABEL. `docs/TODO.md` § 88.1c
    /// spent a whole entry arriving at this: a handle is claimed, a player id is issued, and any
    /// system that keys off the claim can be impersonated by claiming it. **A friendship is
    /// between ids.** The handle is stored beside it only so a rail can draw somebody who is
    /// offline, and it is refreshed from the service whenever they are seen.
    ///
    /// ⚠️ `JoinCode` IS THE WHOLE OF PARTIES IN THIS GAME, and it is not a placeholder. A party
    /// is "be in the same lobby", the game already resolves 4-character join codes LAN-first then
    /// online (`ServerQuery`), and a friend who is in a joinable lobby publishes theirs. There is
    /// no matchmaker to queue into until Phase 7, so a party that "queues together" would be a
    /// button that cannot do anything.
    /// </summary>
    [Serializable]
    public sealed class FriendRef
    {
        public string PlayerId = "";
        public string Handle = "";
        public PresenceState Presence = PresenceState.Offline;

        /// <summary>The lobby they are in, if they are in one that can be joined.</summary>
        public string JoinCode = "";

        /// <summary>ISO-8601 UTC, last time the service saw them.</summary>
        public string SeenUtc = "";
    }

    /// <summary>
    /// One player's whole social state, as one document.
    ///
    /// ⚠️⚠️ ONE DOCUMENT, NOT FOUR, AND `Banner.cs` MADE THE SAME CALL FOR THE SAME REASON. Four
    /// lists that are always read together and always written together are one thing with four
    /// fields; splitting them is four Cloud Save keys, four round trips, and four ways for the
    /// halves to disagree about whether a request that was just accepted is still pending.
    ///
    /// ⚠️ INCOMING AND OUTGOING ARE SEPARATE LISTS RATHER THAN ONE LIST WITH A DIRECTION FLAG.
    /// They are answered by different sides, drawn in different places and acted on with
    /// different verbs: you ACCEPT an incoming and you CANCEL an outgoing. One list with a flag
    /// is a list every reader has to filter, and one that eventually gets filtered wrongly.
    /// </summary>
    [Serializable]
    public sealed class SocialList
    {
        public List<FriendRef> Friends = new List<FriendRef>();
        public List<FriendRef> Incoming = new List<FriendRef>();
        public List<FriendRef> Outgoing = new List<FriendRef>();

        /// <summary>
        /// ⚠️⚠️ IDS ONLY, AND DELIBERATELY NOT `FriendRef`. A block list holding handles would
        /// keep drawing the name of somebody you blocked, which is the one list where showing the
        /// person back to you is the failure. It also has to survive them renaming themselves,
        /// which is exactly what keying on the id buys.
        /// </summary>
        public List<string> Blocked = new List<string>();
    }

    /// <summary>
    /// Friends, blocks and presence, as rules rather than as a screen.
    ///
    /// ⚠️⚠️ THE SERVER RUNS THE SAME DECISIONS AND `ugs/cloud-code/social.js` MIRRORS THEM, for
    /// the reason `FUTURE.md` § 0.5 rule 6 gives and the reason `BannerRules.Normalise` is in the
    /// core: **a client never writes what it owns.** A friend request lands in somebody else's
    /// document, so it is the one thing in this game a peer can put in front of a stranger, and
    /// every cap and every refusal below is what stops that being a way to spam them.
    ///
    /// ⚠️ EVERY METHOD IS PURE AND NONE OF THEM TOUCH A NETWORK. That is what makes a four-way
    /// friend/block interaction assertable in a millisecond instead of with four accounts.
    /// </summary>
    public static class SocialRules
    {
        /// <summary>
        /// ⚠️ 100, AND THE NUMBER IS A DOCUMENT SIZE RATHER THAN A DESIGN STATEMENT. The whole
        /// list round-trips through one Cloud Save record on every load, and a record that grows
        /// without a ceiling is a load that gets slower for ever. At roughly 120 bytes a row this
        /// is about 12 KB, which is the same order as `PlayerProfile`'s match history.
        /// </summary>
        public const int MaxFriends = 100;

        /// <summary>⚠️ THE SAME CEILING, because a griefer's answer to a friend cap is a block
        /// list, and a block list that fills up silently stops protecting anybody.</summary>
        public const int MaxBlocked = 100;

        /// <summary>
        /// ⚠️⚠️ PENDING IS CAPPED LOWER THAN FRIENDS ON PURPOSE, AND IT IS THE ANTI-SPAM RULE.
        /// An incoming request is something a stranger can create in your document. Twenty is
        /// more than anybody will ever have outstanding and few enough that a full inbox is a
        /// visible state rather than a scrolling wall.
        /// </summary>
        public const int MaxPending = 20;

        /// <summary>
        /// ⚠️ HOW LONG A PRESENCE IS BELIEVED. Presence is a written timestamp rather than a live
        /// socket, so "online" is really "wrote a heartbeat recently"; without a staleness bound
        /// a player who closed the game stays lit for ever and the one thing on the screen becomes
        /// a lie. Three heartbeat intervals, so one missed write does not blink somebody out.
        /// </summary>
        public const int PresenceStaleSeconds = 180;

        /// <summary>
        /// ⚠️⚠️ HOW OFTEN A CLIENT MAY WRITE ITS PRESENCE, AND IT IS NOT `ServerQuery`'S 4 s.
        /// `FUTURE.md` § 19.6: *"presence polling must not raise the service query rate."* The
        /// lobby browser queries every 4 seconds because a lobby list goes stale in seconds; a
        /// friends rail does not, and writing presence at that rate would be **fifteen times the
        /// writes for a fact that changes when you press PLAY.** Sixty seconds, and the client
        /// piggybacks on the existing timer rather than adding a second one.
        /// </summary>
        public const int PresenceWriteSeconds = 60;

        /// <summary>
        /// Whether a player id is a plausible target for anything here.
        ///
        /// ⚠️ IT REFUSES THE EMPTY STRING, WHICH IS THE ONE THAT MATTERS. `PlayerAccount.PlayerId`
        /// is empty on a machine that has never reached the service (`docs/TODO.md` § 97), so
        /// without this every offline player would be "friends" with every other offline player,
        /// all of them keyed on "".
        /// </summary>
        public static bool IsAddressable(string playerId)
            => !string.IsNullOrEmpty(playerId) && playerId.Trim().Length > 0;

        public static bool IsBlocked(SocialList list, string playerId)
        {
            if (list?.Blocked == null || !IsAddressable(playerId)) return false;

            foreach (var id in list.Blocked)
                if (id == playerId) return true;

            return false;
        }

        public static bool IsFriend(SocialList list, string playerId)
            => Find(list?.Friends, playerId) != null;

        public static FriendRef Find(List<FriendRef> where, string playerId)
        {
            if (where == null || !IsAddressable(playerId)) return null;

            foreach (var row in where)
                if (row != null && row.PlayerId == playerId) return row;

            return null;
        }

        /// <summary>
        /// Why a request cannot be sent, or empty if it can.
        ///
        /// ⚠️⚠️ IT ANSWERS A SENTENCE RATHER THAN A BOOL, AND THAT IS A UI DECISION MADE IN THE
        /// CORE ON PURPOSE. A disabled ADD button with no reason is `CLAUDE.md` § 6.3's dead
        /// control: the player cannot tell "already sent" from "they blocked you" from "your list
        /// is full", and the three have completely different answers. The string is short, it is
        /// the same on both sides because the server runs this too, and there is exactly one
        /// place to change the wording.
        ///
        /// ⚠️ "THEY BLOCKED YOU" IS NOT ONE OF THE ANSWERS, AND THAT IS DELIBERATE. The sender's
        /// own document cannot see the recipient's block list, and telling somebody they have been
        /// blocked is how a block becomes an argument. The server refuses it silently and the
        /// sender sees a request that simply stays pending, which is what every shipping game
        /// does.
        /// </summary>
        public static string WhyCannotRequest(SocialList list, string me, string them)
        {
            if (!IsAddressable(them)) return "no player to add";
            if (!IsAddressable(me)) return "sign in to add friends";
            if (me == them) return "that is you";

            if (IsBlocked(list, them)) return "unblock them first";
            if (IsFriend(list, them)) return "already friends";
            if (Find(list?.Outgoing, them) != null) return "request already sent";

            // ⚠️ AN INCOMING REQUEST FROM THEM IS NOT A REFUSAL, IT IS THE ANSWER. Two people
            // adding each other at the same time is the commonest race a friends list has, and
            // the polite resolution is that the second press accepts the first request. The
            // caller is told to accept rather than being told no.
            if (Find(list?.Incoming, them) != null) return "";

            if (list != null && list.Friends != null && list.Friends.Count >= MaxFriends)
                return "your friends list is full";

            if (list != null && list.Outgoing != null && list.Outgoing.Count >= MaxPending)
                return "too many requests pending";

            return "";
        }

        public static bool CanRequest(SocialList list, string me, string them)
            => WhyCannotRequest(list, me, them).Length == 0;

        /// <summary>
        /// Whether a request from <paramref name="them"/> may be delivered into this document.
        ///
        /// ⚠️⚠️ THIS IS THE RECIPIENT'S SIDE AND IT IS THE ONE THE SERVER MUST RUN. Everything in
        /// `WhyCannotRequest` is about the sender's own state, which a modified client simply
        /// lies about. **A block is only a block if the refusal happens where the write lands.**
        /// </summary>
        public static bool AcceptsRequestFrom(SocialList list, string them)
        {
            if (!IsAddressable(them)) return false;
            if (IsBlocked(list, them)) return false;
            if (IsFriend(list, them)) return false;
            if (Find(list?.Incoming, them) != null) return false;

            return list?.Incoming == null || list.Incoming.Count < MaxPending;
        }

        /// <summary>
        /// Whether this friend's presence is recent enough to believe.
        ///
        /// ⚠️ A MISSING OR UNPARSEABLE TIMESTAMP IS OFFLINE, NEVER ONLINE. A record written by an
        /// older build, or one whose clock is wrong, must fail towards the quiet answer: a rail
        /// that lights somebody up who is not there sends a player to an empty lobby.
        /// </summary>
        public static bool PresenceIsFresh(FriendRef friend, DateTime utcNow)
        {
            if (friend == null || string.IsNullOrEmpty(friend.SeenUtc)) return false;

            if (!DateTime.TryParse(friend.SeenUtc,
                    System.Globalization.CultureInfo.InvariantCulture,
                    System.Globalization.DateTimeStyles.AdjustToUniversal |
                    System.Globalization.DateTimeStyles.AssumeUniversal, out var seen))
                return false;

            double age = (utcNow - seen).TotalSeconds;
            return age >= 0.0 && age <= PresenceStaleSeconds;
        }

        /// <summary>What this friend is actually doing, after the staleness rule.</summary>
        public static PresenceState EffectivePresence(FriendRef friend, DateTime utcNow)
            => PresenceIsFresh(friend, utcNow) ? friend.Presence : PresenceState.Offline;

        /// <summary>
        /// ⚠️ THE LABEL IS IN THE CORE BECAUSE THE ORDERING DEPENDS ON IT. `Sort` puts joinable
        /// first, then online, then everybody else, and a screen that wrote its own words for
        /// these states could sort by one vocabulary and draw another.
        /// </summary>
        public static string PresenceLabel(PresenceState state)
        {
            switch (state)
            {
                case PresenceState.Menu: return "ONLINE";
                case PresenceState.Queued: return "IN QUEUE";
                case PresenceState.InMatch: return "IN A MATCH";
                case PresenceState.Spectating: return "WATCHING";
                default: return "OFFLINE";
            }
        }

        /// <summary>Whether pressing JOIN on this friend can actually do anything.</summary>
        public static bool IsJoinable(FriendRef friend, DateTime utcNow)
        {
            if (friend == null || string.IsNullOrEmpty(friend.JoinCode)) return false;

            var state = EffectivePresence(friend, utcNow);
            return state != PresenceState.Offline;
        }

        /// <summary>
        /// The rail's order: who you can join, then who is online, then everybody else by name.
        ///
        /// ⚠️⚠️ THIS IS `FUTURE.md` § 0.5b QUESTION 1 AS AN ALGORITHM. The one thing on a friends
        /// list is **who is online now**, and a list sorted alphabetically buries that under
        /// forty offline names. Position is the first of the four ordering tools for a reason:
        /// nothing else on the row has to shout if the row is at the top.
        ///
        /// ⚠️ IT SORTS A COPY. The list it is given is the stored document, and a screen that
        /// reorders storage as a side effect of drawing is a screen that writes when it renders.
        /// </summary>
        public static List<FriendRef> Sorted(List<FriendRef> friends, DateTime utcNow)
        {
            var order = new List<FriendRef>();
            if (friends == null) return order;

            foreach (var row in friends)
                if (row != null && IsAddressable(row.PlayerId)) order.Add(row);

            order.Sort((a, b) =>
            {
                int rankA = Rank(a, utcNow);
                int rankB = Rank(b, utcNow);
                if (rankA != rankB) return rankA.CompareTo(rankB);

                return string.Compare(a.Handle ?? "", b.Handle ?? "",
                                      StringComparison.OrdinalIgnoreCase);
            });

            return order;
        }

        private static int Rank(FriendRef friend, DateTime utcNow)
        {
            if (IsJoinable(friend, utcNow)) return 0;
            return EffectivePresence(friend, utcNow) == PresenceState.Offline ? 2 : 1;
        }

        /// <summary>
        /// Everything a stored document may hold, with everything else dropped.
        ///
        /// ⚠️⚠️ THE SAME JOB `BannerRules.Normalise` DOES AND FOR THE SAME REASON: this document
        /// arrives from a service and from a JSON file on the player's disk, and neither is a
        /// promise. It drops rather than refuses, so one corrupt row does not cost somebody their
        /// whole friends list — which is the griefing argument `Banner.cs` records, one system
        /// over.
        ///
        /// ⚠️ A BLOCKED PLAYER IS REMOVED FROM EVERY OTHER LIST, WHICH IS THE ONE RULE HERE THAT
        /// IS NOT MERELY TIDYING. Blocking somebody you are already friends with has to end the
        /// friendship, or the block is a label rather than a boundary, and a pending request from
        /// somebody you blocked would sit in your inbox for ever.
        /// </summary>
        public static SocialList Normalise(SocialList list)
        {
            var clean = new SocialList();
            if (list == null) return clean;

            if (list.Blocked != null)
                foreach (var id in list.Blocked)
                {
                    if (!IsAddressable(id) || clean.Blocked.Contains(id)) continue;
                    if (clean.Blocked.Count >= MaxBlocked) break;
                    clean.Blocked.Add(id);
                }

            clean.Friends = Trim(list.Friends, clean, MaxFriends);
            clean.Incoming = Trim(list.Incoming, clean, MaxPending);
            clean.Outgoing = Trim(list.Outgoing, clean, MaxPending);

            // ⚠️ A ROW IN TWO LISTS IS A CONTRADICTION AND FRIENDS WINS. It is the state that
            // survives a race between "accept" and a stale copy of the pending list, and the
            // friendlier of the two readings of the same facts.
            clean.Incoming = Without(clean.Incoming, clean.Friends);
            clean.Outgoing = Without(clean.Outgoing, clean.Friends);

            return clean;
        }

        private static List<FriendRef> Trim(List<FriendRef> source, SocialList clean, int cap)
        {
            var kept = new List<FriendRef>();
            if (source == null) return kept;

            foreach (var row in source)
            {
                if (row == null || !IsAddressable(row.PlayerId)) continue;
                if (IsBlocked(clean, row.PlayerId)) continue;
                if (Find(kept, row.PlayerId) != null) continue;
                if (kept.Count >= cap) break;

                kept.Add(new FriendRef
                {
                    PlayerId = row.PlayerId,
                    Handle = OneLine(row.Handle, AccountRules.HandleMax),
                    Presence = row.Presence,
                    JoinCode = OneLine(row.JoinCode, 8),
                    SeenUtc = OneLine(row.SeenUtc, 40),
                });
            }

            return kept;
        }

        private static List<FriendRef> Without(List<FriendRef> source, List<FriendRef> exclude)
        {
            var kept = new List<FriendRef>();

            foreach (var row in source)
                if (Find(exclude, row.PlayerId) == null) kept.Add(row);

            return kept;
        }

        private static string OneLine(string raw, int max)
        {
            if (string.IsNullOrEmpty(raw)) return "";

            var built = new System.Text.StringBuilder(raw.Length);

            foreach (char c in raw)
            {
                if (char.IsControl(c)) continue;
                built.Append(c);
                if (built.Length >= max) break;
            }

            return built.ToString().Trim();
        }

        /// <summary>
        /// The people you just played with, who are not already on your list.
        ///
        /// ⚠️⚠️ THIS IS THE ONLY WAY TO ADD SOMEBODY THAT DOES NOT REQUIRE THEM TO HAND YOU
        /// ANYTHING, AND IT IS THEREFORE THE ONE THAT WILL ACTUALLY GET USED. `FUTURE.md` § 6
        /// calls it *"recent players, with the match they were in, and a one-click add"* and
        /// puts the invite on the end-of-match screen because that is *"the highest-converting
        /// social prompt any game of this shape has"*. Everything the row needs is already in the
        /// `MatchRecord` every peer receives.
        ///
        /// ⚠️ BOTS ARE EXCLUDED BY `IsBot`, NOT BY LOOKING AT THE NAME. `docs/TODO.md` § 94.1
        /// records four lines coming out `IsBot: false` and what believing a name cost.
        ///
        /// ⚠️ AND SO IS THE LOCAL PLAYER, WHICH IS WHY THIS TAKES `me`. The commonest first bug
        /// in a recent-players list is offering to add yourself, and `WhyCannotRequest` would
        /// refuse it correctly — but a row that is drawn and then refused is a control that does
        /// nothing, which `CLAUDE.md` § 6.3 rules out.
        /// </summary>
        public static List<FriendRef> RecentPlayers(MatchRecord record, SocialList list, string me)
        {
            var found = new List<FriendRef>();
            if (record?.Players == null) return found;

            foreach (var line in record.Players)
            {
                if (line == null || line.IsBot) continue;
                if (!IsAddressable(line.PlayerId) || line.PlayerId == me) continue;
                if (IsFriend(list, line.PlayerId) || IsBlocked(list, line.PlayerId)) continue;
                if (Find(list?.Outgoing, line.PlayerId) != null) continue;
                if (Find(found, line.PlayerId) != null) continue;

                found.Add(new FriendRef
                {
                    PlayerId = line.PlayerId,
                    Handle = OneLine(line.Handle, AccountRules.HandleMax),
                    Presence = PresenceState.Menu,
                });
            }

            return found;
        }
    }
}

File: ugs/cloud-code/social.js
const { DataApi } = require("@unity-services/cloud-save-1.4");

// ⚠️⚠️ TWO KEYS PER PLAYER AND NOT ONE. `LIST_KEY` is written by the owner and by anybody who
// sends them a friend request, so it is contended; `PRESENCE_KEY` is written by the owner alone
// every 60 s. Keeping them apart means a heartbeat can never clobber a list, which is exactly
// the race that would otherwise eat a friend request every minute.
const LIST_KEY = "socialList";
const PRESENCE_KEY = "presence";

// ⚠️⚠️ THESE MIRROR `SocialRules` IN THE CORE. This file cannot import a C# constant, so these
// are the places the numbers are written twice; `SocialAndCloudCodeTests` compares them against
// the file on disk and fails if they split. `docs/TODO.md` § 102.
const MAX_FRIENDS = 100;
const MAX_BLOCKED = 100;
const MAX_PENDING = 20;
const PRESENCE_STALE_SECONDS = 180;

// ⚠️⚠️ HOW MANY FRIENDS ONE `load` WILL LOOK UP THE PRESENCE OF. Cloud Save is keyed per player,
// so there is no "read these thirty documents" call: it is thirty reads, server-side, inside one
// invocation the client pays one round trip for. **The cap is what stops a hundred-friend account
// turning one press into a hundred reads**, and the rail draws far fewer than this. Friends past
// it are still listed, they are just drawn from their last known state until they move up.
const PRESENCE_FANOUT = 30;

// ⚠️ MIRRORS `AccountRules.HandleMax` (14 + 1 + 4).
const HANDLE_MAX = 19;

function oneLine(value, max) {
    return String(value || "")
        .replace(/[\u0000-\u001f\u007f]/g, "")
        .trim()
        .slice(0, max);
}

function read(item) {
    if (!item || !item.value) return null;
    try {
        return typeof item.value === "string" ? JSON.parse(item.value) : item.value;
    } catch (e) {
        return null;
    }
}

function emptyList() {
    return { Friends: [], Incoming: [], Outgoing: [], Blocked: [] };
}

function addressable(id) {
    return typeof id === "string" && id.trim().length > 0;
}

function findIn(rows, id) {
    if (!Array.isArray(rows)) return null;
    for (const row of rows) if (row && row.PlayerId === id) return row;
    return null;
}

function without(rows, id) {
    if (!Array.isArray(rows)) return [];
    return rows.filter(row => row && row.PlayerId !== id);
}

function rowFrom(raw) {
    return {
        PlayerId: oneLine(raw && raw.PlayerId, 64),
        Handle: oneLine(raw && raw.Handle, HANDLE_MAX),
        Presence: Number.isInteger(raw && raw.Presence) ? raw.Presence : 0,
        JoinCode: oneLine(raw && raw.JoinCode, 8),
        SeenUtc: oneLine(raw && raw.SeenUtc, 40),
    };
}

// ⚠️⚠️ THIS IS `SocialRules.Normalise` AND IT MUST STAY THE SAME FUNCTION. A document that is
// legal on one side and not on the other is a list that grows on the server and is silently
// trimmed on the client, or the reverse. Blocking removes from every other list here for the
// same reason it does there: a block that leaves a friendship standing is a label, not a
// boundary.
function normalise(raw) {
    const source = raw || {};
    const clean = emptyList();

    for (const id of Array.isArray(source.Blocked) ? source.Blocked : []) {
        const trimmed = oneLine(id, 64);
        if (!addressable(trimmed) || clean.Blocked.includes(trimmed)) continue;
        if (clean.Blocked.length >= MAX_BLOCKED) break;
        clean.Blocked.push(trimmed);
    }

    const trim = (rows, cap) => {
        const kept = [];
        for (const raw of Array.isArray(rows) ? rows : []) {
            const row = rowFrom(raw);
            if (!addressable(row.PlayerId)) continue;
            if (clean.Blocked.includes(row.PlayerId)) continue;
            if (findIn(kept, row.PlayerId)) continue;
            if (kept.length >= cap) break;
            kept.push(row);
        }
        return kept;
    };

    clean.Friends = trim(source.Friends, MAX_FRIENDS);
    clean.Incoming = trim(source.Incoming, MAX_PENDING);
    clean.Outgoing = trim(source.Outgoing, MAX_PENDING);

    // A row in two lists is a contradiction and friends wins. See the C#.
    clean.Incoming = clean.Incoming.filter(row => !findIn(clean.Friends, row.PlayerId));
    clean.Outgoing = clean.Outgoing.filter(row => !findIn(clean.Friends, row.PlayerId));

    return clean;
}

async function loadList(api, projectId, who) {
    const response = await api.getProtectedItems(projectId, who, [LIST_KEY]);
    return normalise(read(response.data.results.find(x => x.key === LIST_KEY)));
}

async function saveList(api, projectId, who, list) {
    await api.setProtectedItem(projectId, who, {
        key: LIST_KEY,
        value: JSON.stringify(normalise(list)),
    });
}

async function loadPresence(api, projectId, who) {
    try {
        const response = await api.getProtectedItems(projectId, who, [PRESENCE_KEY]);
        return read(response.data.results.find(x => x.key === PRESENCE_KEY));
    } catch (e) {
        // ⚠️ A PRESENCE THAT CANNOT BE READ IS OFFLINE, NEVER AN ERROR. One friend's unreadable
        // document must not fail the whole list: the rail would go blank because somebody else
        // has a problem, which is the least explicable failure a social screen can have.
        return null;
    }
}

// ⚠️⚠️ THE SERVICE TOKEN IS WHAT MAKES THIS ENDPOINT POSSIBLE AT ALL, AND IT IS ALSO THE WHOLE
// RISK SURFACE. Every action below that touches another player's document goes through here, and
// `player-account.js`'s `verify` branch carries the same note: a missing service token must read
// as "could not do it" rather than as a silent success against the caller's own document, which
// is what a plain `new DataApi(context)` would do.
function serviceStore(context) {
    try {
        return new DataApi({ accessToken: context.serviceToken });
    } catch (e) {
        throw new Error("social unavailable");
    }
}

module.exports = async ({ params, context, logger }) => {
    const api = new DataApi(context);
    const { projectId, playerId } = context;
    const action = String(params.action || "load");
    const subject = oneLine(params.playerId, 64);
    const handle = oneLine(params.handle, HANDLE_MAX);

    // -----------------------------------------------------------------------
    // LOAD, with the presence of the friends the rail is about to draw.
    // -----------------------------------------------------------------------
    if (action === "load") {
        const mine = await loadList(api, projectId, playerId);
        const store = serviceStore(context);
        const cutoff = Date.now() - PRESENCE_STALE_SECONDS * 1000;

        for (let i = 0; i < mine.Friends.length && i < PRESENCE_FANOUT; i++) {
            const friend = mine.Friends[i];
            const seen = await loadPresence(store, projectId, friend.PlayerId);
            if (!seen) continue;

            const when = Date.parse(String(seen.SeenUtc || ""));

            // ⚠️ THE STALENESS RULE RUNS HERE TOO, so a client on an older build cannot be handed
            // a lit row for somebody who quit an hour ago. The client applies it again; that is
            // not redundancy, it is the same rule at both ends of a document that is also cached
            // to disk (`SocialStore`).
            if (!Number.isFinite(when) || when < cutoff) continue;

            friend.Presence = Number.isInteger(seen.State) ? seen.State : 1;
            friend.JoinCode = oneLine(seen.JoinCode, 8);
            friend.SeenUtc = oneLine(seen.SeenUtc, 40);

            // ⚠️ THE HANDLE IS REFRESHED FROM THE PRESENCE DOCUMENT RATHER THAN LEFT AS IT WAS
            // WHEN THEY WERE ADDED, so a friend who renames themselves is not a stale name on
            // forty other people's screens for ever.
            if (seen.Handle) friend.Handle = oneLine(seen.Handle, HANDLE_MAX);
        }

        return { list: JSON.stringify(mine) };
    }

    // -----------------------------------------------------------------------
    // PRESENCE, which is the only thing written on a timer.
    // -----------------------------------------------------------------------
    if (action === "presence") {
        const state = Number.isInteger(params.state) ? params.state : 1;

        await api.setProtectedItem(projectId, playerId, {
            key: PRESENCE_KEY,
            value: JSON.stringify({
                State: state < 0 || state > 4 ? 1 : state,
                JoinCode: oneLine(params.joinCode, 8),
                Handle: handle,
                SeenUtc: new Date().toISOString(),
            }),
        });

        return { written: true };
    }

    // -----------------------------------------------------------------------
    // REQUEST. The one action that writes into a stranger's document.
    // -----------------------------------------------------------------------
    if (action === "request") {
        if (!addressable(subject) || subject === playerId) throw new Error("no player to add");

        const mine = await loadList(api, projectId, playerId);

        if (mine.Blocked.includes(subject)) throw new Error("unblock them first");
        if (findIn(mine.Friends, subject)) return { list: JSON.stringify(mine) };
        if (mine.Outgoing.length >= MAX_PENDING) throw new Error("too many requests pending");
        if (mine.Friends.length >= MAX_FRIENDS) throw new Error("your friends list is full");

        const store = serviceStore(context);
        const theirs = await loadList(store, projectId, subject);

        // ⚠️⚠️ THEIR BLOCK LIST DECIDES, AND THE CALLER IS NOT TOLD. `SocialRules` carries the
        // reasoning: telling somebody they have been blocked is how a block becomes an argument.
        // The request is dropped and the caller sees a pending row that is never accepted, which
        // is what every shipping game does.
        const refused = theirs.Blocked.includes(playerId) ||
                        theirs.Incoming.length >= MAX_PENDING;

        // ⚠️ THE MUTUAL CASE IS AN ACCEPT, NOT A SECOND REQUEST. Two people adding each other at
        // once is the commonest race a friends list has, and this is where it resolves.
        if (findIn(mine.Incoming, subject)) {
            // ⚠️ THE HANDLE COMES OFF THE ROW THEY CREATED, NOT OFF THE CALLER'S PARAMETERS.
            // They wrote their own name into our inbox when they sent the request, which is the
            // only value here that they own; `params.theirHandle` is what WE last saw and would
            // be empty on the mutual path, because nobody has drawn them yet.
            const pending = findIn(mine.Incoming, subject);
            mine.Incoming = without(mine.Incoming, subject);
            mine.Friends.push(pending);

            theirs.Outgoing = without(theirs.Outgoing, playerId);
            theirs.Friends.push(rowFrom({ PlayerId: playerId, Handle: handle }));

            await saveList(store, projectId, subject, theirs);
            await saveList(api, projectId, playerId, mine);
            return { list: JSON.stringify(normalise(mine)) };
        }

        if (!findIn(mine.Outgoing, subject))
            mine.Outgoing.push(rowFrom({ PlayerId: subject, Handle: oneLine(params.theirHandle, HANDLE_MAX) }));

        await saveList(api, projectId, playerId, mine);

        if (!refused && !findIn(theirs.Incoming, playerId) && !findIn(theirs.Friends, playerId)) {
            theirs.Incoming.push(rowFrom({ PlayerId: playerId, Handle: handle }));
            await saveList(store, projectId, subject, theirs);
        }

        return { list: JSON.stringify(normalise(mine)) };
    }

    // -----------------------------------------------------------------------
    // ACCEPT and DECLINE, which act on a row somebody else created.
    // -----------------------------------------------------------------------
    if (action === "accept" || action === "decline") {
        if (!addressable(subject)) throw new Error("no player named");

        const mine = await loadList(api, projectId, playerId);
        const pending = findIn(mine.Incoming, subject);
        if (!pending) return { list: JSON.stringify(mine) };

        mine.Incoming = without(mine.Incoming, subject);

        const store = serviceStore(context);
        const theirs = await loadList(store, projectId, subject);
        theirs.Outgoing = without(theirs.Outgoing, playerId);

        if (action === "accept") {
            if (!findIn(mine.Friends, subject)) mine.Friends.push(pending);
            if (!findIn(theirs.Friends, playerId))
                theirs.Friends.push(rowFrom({ PlayerId: playerId, Handle: handle }));
        }

        await saveList(store, projectId, subject, theirs);
        await saveList(api, projectId, playerId, mine);

        return { list: JSON.stringify(normalise(mine)) };
    }

    // -----------------------------------------------------------------------
    // REMOVE and BLOCK. Both end a friendship on BOTH sides.
    // -----------------------------------------------------------------------
    if (action === "remove" || action === "block") {
        if (!addressable(subject) || subject === playerId) throw new Error("no player named");

        const mine = await loadList(api, projectId, playerId);

        mine.Friends = without(mine.Friends, subject);
        mine.Incoming = without(mine.Incoming, subject);
        mine.Outgoing = without(mine.Outgoing, subject);

        if (action === "block" && !mine.Blocked.includes(subject)) {
            if (mine.Blocked.length >= MAX_BLOCKED) throw new Error("your block list is full");
            mine.Blocked.push(subject);
        }

        // ⚠️⚠️ THE OTHER SIDE IS CLEARED TOO, AND SKIPPING IT WOULD LEAVE A ONE-WAY FRIENDSHIP.
        // They would keep drawing your presence, keep offering JOIN, and keep being told you are
        // online, from a list you are no longer on. **A friendship is one fact stored twice**, so
        // both copies move together or neither does.
        const store = serviceStore(context);
        const theirs = await loadList(store, projectId, subject);

        theirs.Friends = without(theirs.Friends, playerId);
        theirs.Incoming = without(theirs.Incoming, playerId);
        theirs.Outgoing = without(theirs.Outgoing, playerId);

        await saveList(store, projectId, subject, theirs);
        await saveList(api, projectId, playerId, mine);

        return { list: JSON.stringify(normalise(mine)) };
    }

    if (action === "unblock") {
        const mine = await loadList(api, projectId, playerId);
        mine.Blocked = mine.Blocked.filter(id => id !== subject);
        await saveList(api, projectId, playerId, mine);
        return { list: JSON.stringify(normalise(mine)) };
    }

    throw new Error("unknown social action");
};

// ⚠️⚠️ EVERY PARAMETER A SCRIPT USES MUST BE DECLARED HERE OR CLOUD CODE STRIPS IT, AND THE
// FAILURE IS SILENT AND LOOKS LIKE A WORKING ENDPOINT. `docs/TODO.md` § 90.5: three scripts
// declared none, so every call fell through to whichever branch `params.action || "default"`
// picked, answered normally, and no career had ever reached the server. Every probe was green
// throughout, because they all probed with the default branch.
//
// ⚠️ TWO FURTHER THINGS SILENTLY DROP THE WHOLE DECLARATION AND PRINT `params: []`: a parameter
// typed `JSON` (use `String` and serialise), and a top-level `function parameters()` anywhere in
// the file. `ugs cloud-code scripts get social` is the only way to see what the service holds.
//
// ⚠️ IF A NEW ACTION NEEDS A NEW PARAMETER, IT GOES IN THIS BLOCK IN THE SAME EDIT.
module.exports.params = {
    action: "String",

    // The other player, for every action that names one.
    playerId: "String",

    // The CALLER's own handle, so the row that lands in somebody else's document has a name on
    // it. ⚠️ It is a label and never an identity: the id is what the friendship is keyed on.
    handle: "String",

    // The SUBJECT's handle as the caller last saw it, so an outgoing row can be drawn before
    // they have ever been seen online.
    theirHandle: "String",

    // `presence` only.
    state: "Numeric",
    joinCode: "String",
};
